### PR TITLE
Tctti get poll handles

### PR DESCRIPTION
**Pull request by williamcroberts**
### Added
- tcti-mssim now supports Tss2\_Tcti\_GetPollHandles()
- tcti-swtpm now supports Tss2\_Tcti\_GetPollHandles()

### CODE REVIEW

**Comment by williamcroberts** [comment]
@pdxjohnny here you go

**Comment by williamcroberts** [comment]
So it seems to be failing on windows because, AFAICT, it's building, the async stuff and the structure isn't defined... But thats the else clause I would expect it to hit the _WIN32 define

**Comment by williamcroberts** [comment]
Fapi tests seem to be hanging, maybe I need that fapi async magic.. not sure

**Comment by williamcroberts** [comment]
Windows is failing:
```
  tcti-mssim.c(286,16): error : member reference base type 'TSS2_TCTI_POLL_HANDLE' (aka 'void *') is not a structure or union [C:\projects\tpm2-tss\src\tss2-tcti\tss2-tcti-mssim.vcxproj]
```
But I would imagine it should be defining it as HANDLE not VOID

**Comment by williamcroberts** [comment]
hrmm, added some debug logs:
```
trace:fapi:src/tss2-fapi/fapi_util.c:963:ifapi_session_init() call 
trace:fapi:src/tss2-fapi/fapi_util.c:1166:ifapi_get_sessions_async() call 
trace:fapi:src/tss2-fapi/ifapi_keystore.c:546:ifapi_keystore_load_async() Load object: P_ECC/HS/SRK 
trace:fapi:src/tss2-fapi/api/Fapi_GetRandom.c:165:Fapi_GetRandom_Async() finished 
trace:fapi:src/tss2-fapi/api/Fapi_GetPollHandles.c:51:Fapi_GetPollHandles() called for context:0x564c094b3d40 
debug:fapi:src/tss2-fapi/ifapi_io.c:685:ifapi_io_poll_handles() No pollable operation in progress. 
trace:tcti:src/tss2-tcti/tcti-mssim.c:274:tcti_mssim_get_poll_handles() tcti_swtpm_get_poll_handles getting poll handle 
trace:tcti:src/tss2-tcti/tcti-mssim.c:274:tcti_mssim_get_poll_handles() tcti_swtpm_get_poll_handles getting poll handle 
debug:fapi:src/tss2-fapi/api/Fapi_GetPollHandles.c:87:Fapi_GetPollHandles() Returning 1 ESYS poll handles. 
trace:fapi:src/tss2-fapi/api/Fapi_GetPollHandles.c:88:Fapi_GetPollHandles() finished 
```

It looks like it's getting the tpm_sock but nothing is coming back...

**Comment by AndreasFuchsSIT** [comment]
> It looks like it's getting the tpm_sock but nothing is coming back...

You need to set the signal mask: `handle->events = POLLIN` is my guess.

**Comment by williamcroberts** [comment]
> > It looks like it's getting the tpm_sock but nothing is coming back...
> 
> You need to set the event mask: `handle->events = POLLIN` is my guess.

So the tcti-device.c and tcti-swtpm.c implementations are not doing that. How well tested is this async code?

So we need to pick up the delta on windoze builds where they just populate an array of HANDLES... is that just the FD? Or do we just return not supported for windoze?

**Comment by tstruk** [comment]
> > It looks like it's getting the tpm_sock but nothing is coming back...
> 
> You need to set the event mask: `handle->events = POLLIN` is my guess.

This should be set by the application, depending on what events it want to get, and not the tcti.

File: src/tss2-tcti/tcti-mssim.c [hunk 1]
```diff
@@ -271,10 +271,22 @@ tcti_mssim_get_poll_handles (
...
-    (void)(num_handles);
-    return TSS2_TCTI_RC_NOT_IMPLEMENTED;
+    TSS2_TCTI_MSSIM_CONTEXT *tcti_mssim = tcti_mssim_context_cast (tctiContext);
+
+    if (num_handles == NULL || tcti_mssim == NULL) {
+        return TSS2_TCTI_RC_BAD_REFERENCE;
+    }
+
+    if (handles != NULL && *num_handles < 1) {
+        return TSS2_TCTI_RC_INSUFFICIENT_BUFFER;
```
Comment (tstruk) [inline src/tss2-tcti/tcti-mssim.c:281]: INSUFFICIENT_BUFFER is mainly for the command/response buffer. This should be BAD_VALUE

**Comment by williamcroberts** [comment]
> > > It looks like it's getting the tpm_sock but nothing is coming back...
> > 
> > 
> > You need to set the event mask: `handle->events = POLLIN` is my guess.
> 
> This should be set by the application, depending on what events it want to get, and not the tcti.

Ok thats what I thought initially

**Comment by tstruk** [comment]
I spent some time trying to debug the issues with async on sw emulators.
The changes can be found here:
https://github.com/tstruk/tpm2-tss/commit/ea68f8c857b9cd6a96cf48f8306d419802ee7811
https://github.com/tstruk/tpm2-tss/commit/2e96cfd962f3dc54e0d3d843d4823b3c5101938d

The status is that it is working fine with mssim (./configure --disable-tcti-swtpm) i.e. all integration tests are passing.
It is still failing with the swtpm simulator. Swtpm_tcti design is different to mssim. For whatever reason it opens a new socket on every transmit() and closes it on every receive(), which makes the poll handle invalid after every command/response. When I changed this to the same design that we have in mssim, i.e. create the socket at swtpm_init(), make it nonbloking, and use it until swtpm_finalize(), it behaves weird. It works for the first few commands/responses and basically returns rubbish in subsequent calls. When it reads the response header and tries to get the actual size of the response it gets rubbish and fails as shown below (**tcti_swtpm_receive() response size: 369098752**):

```
debug:tcti:src/tss2-tcti/tcti-swtpm.c:275:tcti_swtpm_transmit() Sending command with TPM_CC 0x17a and size 22 
debug:tcti:src/util/io.c:136:socket_xmit_buf() Writing 22 bytes to socket 3: (size=22): 
0000: 8001000000160000017a000000010000  .........z......
0010: 0000000000fe                      ......
debug:tcti:src/util/io.c:90:write_all() writing 22 bytes starting at 0x6a3ca0 to fd 3 
debug:tcti:src/util/io.c:104:write_all() wrote 22 bytes to fd 3 
debug:tcti:src/tss2-tcti/tcti-swtpm.c:409:tcti_swtpm_receive() Receiving header to determine the size of the response. 
debug:tcti:src/tss2-tcti/tcti-swtpm.c:437:tcti_swtpm_receive() response size: 369098752
ERROR:test:test/helper/tpm_dumpstate.c:67:main() TPM GetCapabilities FAILED: 0x80012 
Error during dumpstate
```

I presume that's the reason why swtpm needs to open a new connection for every command. @joholl can you confirm that? If that's the case then we will not be able to support async for swtpm.

**Comment by tstruk** [comment]
Here is a PR that enables async mssim.

**Comment by williamcroberts** [comment]
I forgot about that open-close sequence. It's not impossible to support, I see some avenues forward,
But I don't think the work is worth the reward with the thoughts I have in mind (they involve another thread
and a pipe).

**Comment by williamcroberts** [comment]
Superseded by #1793

**Pull request closed without merging**